File: core/src/apps/ethereum/sign_tx.py
from micropython import const
from typing import TYPE_CHECKING

from trezor import TR
from trezor.crypto import rlp
from trezor.messages import EthereumTxRequest
from trezor.wire import DataError

from .helpers import (
    address_from_bytes,
    bytes_from_address,
    get_data_confirmer,
    get_progress_indicator,
)
from .keychain import with_keychain_from_chain_id

if TYPE_CHECKING:
    from buffer_types import AnyBytes
    from typing import Sequence

    from trezor.messages import EthereumSignTx
    from trezor.ui.layouts import StrPropertyType
    from trezor.utils import HashWriter

    from apps.common.keychain import Keychain
    from apps.common.payment_request import PaymentRequestVerifier

    from .definitions import Definitions
    from .helpers import ConfirmDataFn, DataChunkLoader
    from .keychain import MsgInSignTx


# Maximum chain_id which returns the full signature_v (which must fit into an uint32).
# chain_ids larger than this will only return one bit and the caller must recalculate
# the full value: v = 2 * chain_id + 35 + v_bit
_MAX_CHAIN_ID = const(0xFFFF_FFFF - 36) // 2


@with_keychain_from_chain_id
async def sign_tx(
    msg: EthereumSignTx,
    keychain: Keychain,
    defs: Definitions,
) -> EthereumTxRequest:
    from trezor.ui.layouts import show_continue_in_app

    from apps.common import paths

    from .helpers import format_ethereum_amount, get_fee_items_regular, keccak256

    # local_cache_attribute
    data_length = msg.data_length
    tx_type = msg.tx_type
    network = defs.network

    check_common_fields(msg)

    address_bytes = bytes_from_address(msg.to)

    valid_tx_types = (1, 6, None)
    if tx_type not in valid_tx_types:
        raise DataError("tx_type out of bounds")
    if len(msg.gas_price) + len(msg.gas_limit) > 30:
        raise DataError("Fee overflow")

    # have the user confirm signing
    await paths.validate_path(keychain, msg.address_n)
    sender_bytes = keychain.derive(msg.address_n).ethereum_pubkeyhash()
    gas_price = int.from_bytes(msg.gas_price, "big")
    gas_limit = int.from_bytes(msg.gas_limit, "big")
    maximum_fee = format_ethereum_amount(gas_price * gas_limit, None, network)
    fee_items = get_fee_items_regular(
        gas_price,
        gas_limit,
        network,
    )

    payment_req_verifier = None
    if msg.payment_req:
        from apps.common.payment_request import PaymentRequestVerifier

        slip44_id = paths.unharden(msg.address_n[1])
        payment_req_verifier = PaymentRequestVerifier(
            msg.payment_req,
            slip44_id,
            keychain,
            amount_size_bytes=32,
        )

    sha = keccak256()
    rlp.write_header(sha, _get_digest_length(msg, data_length), rlp.LIST_HEADER_BYTE)

    if tx_type is not None:
        rlp.write(sha, tx_type)

    for field in (msg.nonce, msg.gas_price, msg.gas_limit, address_bytes, msg.value):
        rlp.write(sha, field)

    initial_data = await request_initial_data(msg, sha)

    # Confirm the transaction, using special layouts for staking, yielding and clear-signing (if supported).
    await confirm_tx_data(
        initial_data,
        msg,
        defs,
        address_bytes,
        maximum_fee,
        fee_items,
        payment_req_verifier,
        sender_bytes,
        # Hash and confirm the rest of transaction calldata while loading it from the host.
        create_data_chunk_loader(sha),
    )

    # eip 155 replay protection
    rlp.write(sha, msg.chain_id)
    rlp.write(sha, 0)
    rlp.write(sha, 0)

    digest = sha.get_digest()

    # transaction data confirmed, proceed with signing
    result = _sign_digest(msg, keychain, digest)

    show_continue_in_app(TR.send__transaction_signed)
    return result


_MAX_DATA_STORED = const(6144)
_DATA_CHUNK_SIZE = const(1024)


async def request_initial_data(msg: MsgInSignTx, sha: HashWriter) -> AnyBytes:
    """Request at most `MAX_DATA_STORED` which we keep locally"""
    from trezor.utils import empty_bytearray

    data_length = msg.data_length
    if data_length > len(msg.data_initial_chunk):
        # pre-allocate memory
        buf_capacity = min(data_length, _MAX_DATA_STORED)
        buf = empty_bytearray(buf_capacity)
        buf.extend(msg.data_initial_chunk)

        # preload `buf_capacity` bytes from the host into `buf`
        while (data_left := buf_capacity - len(buf)) > 0:
            chunk = await _get_next_chunk(data_left)
            buf.extend(chunk)
    else:
        buf = msg.data_initial_chunk

    rlp.write_header(sha, data_length, rlp.STRING_HEADER_BYTE, buf)
    sha.extend(buf)
    return buf


async def confirm_tx_data(
    initial_data: AnyBytes,
    msg: MsgInSignTx,
    defs: Definitions,
    address_bytes: bytes,
    maximum_fee: str,
    fee_items: Sequence[StrPropertyType],
    payment_request_verifier: PaymentRequestVerifier | None,
    sender_bytes: AnyBytes,
    data_chunk_loader: DataChunkLoader,
) -> None:
    """Clear-sign the transaction or confirm calldata chunks."""

    from . import clear_signing, staking, yielding
    from .helpers import format_ethereum_amount
    from .layout import require_confirm_payment_request, require_confirm_tx

    # local_cache_attribute
    data_length = msg.data_length
    network = defs.network

    staking_approver = staking.get_approver(
        msg, network, address_bytes, maximum_fee, fee_items
    )
    if staking_approver is not None:
        if payment_request_verifier is not None:
            raise DataError("Payment Requests don't support staking")
        return await staking_approver

    yielding_approver = await yielding.get_approver(
        msg, initial_data, network, address_bytes, maximum_fee, fee_items, sender_bytes
    )
    if yielding_approver is not None:
        if payment_request_verifier is not None:
            raise DataError("Payment Requests don't support yielding")
        return await yielding_approver

    value = int.from_bytes(msg.value, "big")

    if len(initial_data) < data_length:
        # Don't even attempt to clear sign if we have a calldata larger than `MAX_DATA_STORED`.
        # this is because clear signing doesn't currently support fetching additional data,
        # which is because if it did, we would not be able to fall back to blind signing anymore.
        clear_signed = False
    else:
        try:
            clear_signed = await clear_signing.try_confirm(
                initial_data,
                address_bytes,
                msg,
                defs,
                maximum_fee,
                fee_items,
                payment_request_verifier,
            )
        except clear_signing.ClearSigningFailed:
            clear_signed = False

    recipient_str = (
        address_from_bytes(address_bytes, network) if address_bytes else None
    )

    if payment_request_verifier is not None:
        if data_length != 0:
            raise DataError(
                "Data length must be 0 when `payment_request_verifier` is provided."
            )

        if msg.payment_req is None:
            raise DataError(
                "Payment request (`payment_req`) must not be None when `payment_request_verifier` is provided."
            )

        assert recipient_str is not None

        payment_request_verifier.add_output(value, recipient_str or "")
        payment_request_verifier.verify()
        return await require_confirm_payment_request(
            recipient_str,
            msg.payment_req,
            msg.address_n,
            maximum_fee,
            fee_items,
            msg.chain_id,
            network,
            None,
            None,
        )
    elif not clear_signed:
        if data_length > 0:
            confirm_data_chunk = get_data_confirmer(data_length)
        else:
            confirm_data_chunk = get_progress_indicator(data_length)
        token = (
            None  # what we want to confirm here is the ETH amount being sent on-chain
        )

        # Stream, confirm and hash the rest of the calldata chunks.
        await _confirm_data_chunks(
            confirm_data_chunk, initial_data, data_length, data_chunk_loader
        )
        return await require_confirm_tx(
            recipient_str,
            format_ethereum_amount(value, token, network),
            address_bytes,
            msg.address_n,
            maximum_fee,
            fee_items,
            token,
            is_send=(data_length == 0),
            chunkify=bool(msg.chunkify),
        )


def _get_digest_length(msg: EthereumSignTx, data_total: int) -> int:
    length = 0
    if msg.tx_type is not None:
        length += rlp.length(msg.tx_type)

    fields: tuple[rlp.RLPItem, ...] = (
        msg.nonce,
        msg.gas_price,
        msg.gas_limit,
        bytes_from_address(msg.to),
        msg.value,
        msg.chain_id,
        0,
        0,
    )

    for field in fields:
        length += rlp.length(field)

    length += rlp.header_length(data_total, msg.data_initial_chunk)
    length += data_total

    return length


def create_data_chunk_loader(h: HashWriter) -> DataChunkLoader:
    async def data_chunk_loader(data_left: int) -> AnyBytes:
        chunk = await _get_next_chunk(data_left)
        h.extend(chunk)
        return chunk

    return data_chunk_loader


async def _confirm_data_chunks(
    confirm_data_chunk: ConfirmDataFn,
    initial_data: AnyBytes,
    data_length: int,
    data_chunk_loader: DataChunkLoader,
) -> None:
    await confirm_data_chunk(initial_data)
    data_left = data_length - len(initial_data)
    while data_left > 0:
        chunk = await data_chunk_loader(data_left)
        # `confirm_data_chunk` will raise on cancellation, so
        # `data_chunk_loader`-computed hash will be discarded.
        await confirm_data_chunk(chunk)
        data_left -= len(chunk)


async def _get_next_chunk(data_left: int) -> AnyBytes:
    from trezor.messages import EthereumTxAck
    from trezor.wire.context import call

    req = EthereumTxRequest()
    req.data_length = min(data_left, _DATA_CHUNK_SIZE)
    resp = await call(req, EthereumTxAck)
    data_chunk = resp.data_chunk
    if len(data_chunk) != req.data_length:
        raise DataError("Data length mismatch")
    return data_chunk


def _sign_digest(
    msg: EthereumSignTx, keychain: Keychain, digest: bytes
) -> EthereumTxRequest:
    from trezor.crypto.curve import secp256k1

    node = keychain.derive(msg.address_n)
    signature = secp256k1.sign(
        node.private_key(), digest, False, secp256k1.CANONICAL_SIG_ETHEREUM
    )

    req = EthereumTxRequest()
    req.signature_v = signature[0]
    if msg.chain_id > _MAX_CHAIN_ID:
        req.signature_v -= 27
    else:
        req.signature_v += 2 * msg.chain_id + 8

    req.signature_r = signature[1:33]
    req.signature_s = signature[33:]

    return req


def check_common_fields(msg: MsgInSignTx) -> None:
    data_length = msg.data_length  # local_cache_attribute

    if data_length > 0:
        if not msg.data_initial_chunk:
            raise DataError("Data length provided, but no initial chunk")
        # Our encoding only supports transactions up to 2^24 bytes. To
        # prevent exceeding the limit we use a stricter limit on data length.
        if data_length > 16_000_000:
            raise DataError("Data length exceeds limit")
        if len(msg.data_initial_chunk) > data_length:
            raise DataError("Invalid size of initial chunk")

    if len(msg.to) not in (0, 40, 42):
        raise DataError("Invalid recipient address")

    if not msg.to and data_length == 0:
        # sending transaction to address 0 (contract creation) without a data field
        raise DataError("Contract creation without data")

    if msg.chain_id == 0:
        raise DataError("Chain ID out of bounds")

File: sdk/apps/ethereum/src/sign_tx.rs
use crate::{
    alloc_types::{Vec, vec},
    common::{
        EIP_7702_TX_TYPE, check_common_fields, confirm_data_and_summary, confirm_tx_data,
        get_eip_7702_known_address, get_eth_pubkey_hash, request_initial_data,
    },
    definitions::Definitions,
    helpers::{bytes_from_address, format_ethereum_amount, get_fee_items_regular},
    paths::{Bip32Path, unharden},
    payment_request::PaymentRequestVerifier,
    proto::{
        common::button_request::ButtonRequestType,
        ethereum::{SignTx, TxRequest},
    },
    rlp::{self, RLPItem},
};
use primitive_types::U256;
use trezor_app_sdk::{
    Error, Result, ResultExt, crypto,
    ui::{self, Property},
};

// Maximum chain_id which returns the full signature_v (which must fit into an uint32).
// chain_ids larger than this will only return one bit and the caller must recalculate
// the full value: v = 2 * chain_id + 35 + v_bit
const MAX_CHAIN_ID: u64 = (0xFFFF_FFFF - 36) / 2;

pub fn sign_tx(mut msg: SignTx) -> Result<TxRequest> {
    msg.nonce.get_or_insert_default();
    msg.to.get_or_insert_default();
    msg.value.get_or_insert_default();
    msg.data_initial_chunk.get_or_insert_default();
    msg.data_length.get_or_insert_default();

    let nonce = msg
        .nonce
        .as_deref()
        .ok_or(Error::DataError("Missing nonce"))?;
    let to = msg
        .to
        .as_deref()
        .ok_or(Error::DataError("Missing to address"))?;
    let value = msg
        .value
        .as_deref()
        .ok_or(Error::DataError("Missing value"))?;
    let data_initial_chunk = msg
        .data_initial_chunk
        .as_deref()
        .ok_or(Error::DataError("Missing data initial chunk"))?;
    let data_length = msg
        .data_length
        .ok_or(Error::DataError("Missing data length"))?;

    let dp = Bip32Path::from_slice(&msg.address_n);

    let (encoded_network, encoded_token) = match &msg.definitions {
        Some(def) => (def.encoded_network.as_deref(), def.encoded_token.as_deref()),
        None => (None, None),
    };

    let definitions =
        Definitions::from_encoded(encoded_network, encoded_token, Some(msg.chain_id), None)
            .context("Failed to decode definitions")?;

    check_common_fields(data_length, data_initial_chunk, to, msg.chain_id)
        .context("Failed to check common fields")?;

    let address_bytes = bytes_from_address(to).context("Failed to convert address to bytes")?;

    let valid_tx_types = [Some(1), Some(6), Some(EIP_7702_TX_TYPE), None];

    if !valid_tx_types.contains(&msg.tx_type) {
        return Err(Error::DataError("tx_type out of bounds"));
    }

    if matches!(msg.tx_type, Some(EIP_7702_TX_TYPE)) {
        // TODO: safety checks
        // if safety_checks.is_strict():
        // raise DataError("EIP-7702 not allowed in strict checks")

        if get_eip_7702_known_address(&address_bytes).is_none() {
            return Err(Error::DataError("Unknown EIP-7702 address"));
        }
    }

    if msg.gas_price.len() + msg.gas_limit.len() > 30 {
        return Err(Error::DataError("Fee overflow"));
    }

    // have the user confirm signing
    crypto::verify_derivation_path(
        dp.as_slice(),
        encoded_network,
        encoded_token,
        Some(msg.chain_id),
    )
    .context("Failed to verify derivation path")?;

    let sender_bytes = get_eth_pubkey_hash(&dp).context("Failed to get eth pubkey hash")?;

    let gas_price = U256::from_big_endian(&msg.gas_price);
    let gas_limit = U256::from_big_endian(&msg.gas_limit);
    let maximum_fee =
        format_ethereum_amount(gas_price * gas_limit, None, definitions.network(), false)
            .context("Failed to create max fee string")?;
    let fee_items = get_fee_items_regular(gas_price, gas_limit, definitions.network())
        .context("Failed to create fee items")?;
    let fee_items_ref: Vec<Property> = fee_items
        .iter()
        .map(|(k, v, mono)| Property::new(k, v, *mono))
        .collect();

    let payment_req_verifier = if let Some(req) = &msg.payment_req {
        let slip44_id = unharden(dp.as_slice()[1]);
        Some(PaymentRequestVerifier::new(req, slip44_id)?)
    } else {
        None
    };

    let mut hasher = crypto::Keccak256::new(None);
    rlp::write_header(
        &mut hasher,
        get_total_length(
            nonce,
            &msg.gas_price,
            &msg.gas_limit,
            to,
            value,
            data_initial_chunk,
            data_length,
            msg.chain_id,
            msg.tx_type,
        )?,
        rlp::LIST_HEADER_BYTE,
        None,
    );

    if let Some(tx_type) = msg.tx_type {
        rlp::write(&mut hasher, &RLPItem::Int(tx_type.into()));
    }

    // Write transaction fields
    let fields = [
        RLPItem::Bytes(nonce),
        RLPItem::Bytes(&msg.gas_price),
        RLPItem::Bytes(&msg.gas_limit),
        RLPItem::Bytes(&address_bytes),
        RLPItem::Bytes(value),
    ];

    for field in &fields {
        rlp::write(&mut hasher, field);
    }

    let initial_data = request_initial_data(&mut hasher, data_length as usize, data_initial_chunk)
        .context("Failed to request initial data")?;

    let (confirm_data_chunk, confirm_summary) = confirm_tx_data(
        &initial_data,
        data_length as usize,
        data_initial_chunk,
        value,
        to,
        &dp,
        &definitions,
        msg.tx_type,
        &address_bytes,
        &maximum_fee,
        &fee_items_ref,
        payment_req_verifier,
        &sender_bytes,
        msg.chunkify,
        msg.payment_req.as_ref(),
        msg.chain_id,
        msg.definitions.as_ref(),
        msg.supports_definition_request,
    )
    .context("Failed to confirm transaction data")?;

    confirm_data_and_summary(
        confirm_data_chunk,
        confirm_summary,
        &initial_data,
        data_length as usize,
        &mut hasher,
    )
    .context("Failed to confirm data and summary")?;

    // EIP-155 replay protection
    let fields = [
        RLPItem::Int(msg.chain_id.into()),
        RLPItem::Int(0.into()),
        RLPItem::Int(0.into()),
    ];
    for field in &fields {
        rlp::write(&mut hasher, field);
    }

    let digest = hasher.digest();

    // transaction data confirmed, proceed with signing
    let res = sign_digest(&msg, &digest).context("Failed to sign digest")?;
    ui::show_success(ui::ShowSuccess::new(
        tr!("words__title_done"),
        tr!("send__transaction_signed"),
        tr!("instructions__continue_in_app"),
        Some(3200),
        None,
        ButtonRequestType::Other.into(),
    ))?;

    Ok(res)
}

fn sign_digest(msg: &SignTx, digest: &[u8; 32]) -> Result<TxRequest> {
    let (encoded_network, encoded_token) = match &msg.definitions {
        Some(def) => (
            def.encoded_network.as_ref().map(|d| d.as_ref()),
            def.encoded_token.as_ref().map(|d| d.as_ref()),
        ),
        None => (None, None),
    };

    let signature = crypto::sign_typed_hash(
        &msg.address_n,
        digest,
        encoded_network,
        encoded_token,
        Some(msg.chain_id),
        true,
    )
    .context("Failed to sign typed hash")?;

    let mut req = TxRequest::default();
    let mut signature_v: u32 = signature[0].into();

    if msg.chain_id > MAX_CHAIN_ID {
        signature_v -= 27;
    } else {
        signature_v += 2 * msg.chain_id as u32 + 8;
    }

    req.signature_v = Some(signature_v);
    req.signature_r = Some(signature[1..33].to_vec());
    req.signature_s = Some(signature[33..].to_vec());

    Ok(req)
}

pub fn get_total_length(
    nonce: &[u8],
    gas_price: &[u8],
    gas_limit: &[u8],
    to: &str,
    value: &[u8],
    data_initial_chunk: &[u8],
    data_length: u32,
    chain_id: u64,
    tx_type: Option<u32>,
) -> Result<u32> {
    let mut length = 0;

    if let Some(tx_type) = tx_type {
        length += rlp::length(&RLPItem::Int(tx_type.into()));
    }

    let to_bytes = bytes_from_address(to).context("Failed to convert address to bytes")?;

    let fields: Vec<RLPItem> = vec![
        RLPItem::Bytes(nonce),
        RLPItem::Bytes(gas_price),
        RLPItem::Bytes(gas_limit),
        RLPItem::Bytes(&to_bytes),
        RLPItem::Bytes(value),
        RLPItem::Int(chain_id.into()),
        RLPItem::Int(0.into()),
        RLPItem::Int(0.into()),
    ];

    for field in &fields {
        length += rlp::length(field);
    }

    length += rlp::header_length(data_length, Some(data_initial_chunk));
    length += data_length;

    Ok(length)
}
